mbedtls/src/error.rs

```diff
@@ -31,56 +31,98 @@ pub const ERR_UTF8_INVALID: c_int = -0x10000;
-                    _ => -self,
-                };
-                let (high_level_code, low_level_code) = (err_code & 0xFF80, err_code & 0x7F);
-                Err($n::from_mbedtls_code(if high_level_code > 0 { -high_level_code } else { -low_level_code }))
+        impl From<c_int> for $n {
+            fn from(code: c_int) -> $n {
+                match -code {
+                    $(::mbedtls_sys::$c => return $n::$rust),*,
+                    _ => return $n::Unknown(code)
+                }
             }
         }
 
-        impl $n {
-            pub fn from_mbedtls_code(code: c_int) -> Self {
-                match code {
-                    $(::mbedtls_sys::$c => $n::$rust),*,
-                    _ => $n::Other(code)
+        impl From<&$n> for c_int {
+            fn from(error: &$n) -> c_int {
+                match error {
+                    $($n::$rust => return ::mbedtls_sys::$c,)*
+                    $n::Unknown(code) => return *code,
                 }
             }
+        }
 
-            pub fn as_str(&self) -> &'static str {
+        impl $n {
+            pub fn as_str(&self)-> &'static str {
                 match self {
-                    $(&$n::$rust => concat!("mbedTLS error ",stringify!($n::$rust)),)*
-                    &$n::Other(_) => "mbedTLS unknown error",
-                    &$n::Utf8Error(_) => "error converting to UTF-8",
-                    &$n::__Nonexhaustive => unreachable!("__Nonexhaustive value should not be instantiated"),
+                    $($n::$rust => concat!("mbedTLS $n error ", stringify!($n::$rust)),)*
+                    $n::Unknown(_) => "mbedTLS unknown $n error"
                 }
             }
+        }
+    };
+}
 
-            pub fn to_int(&self) -> c_int {
-                match *self {
-                    $($n::$rust => ::mbedtls_sys::$c,)*
-                    $n::Other(code) => code,
-                    $n::Utf8Error(_) => ERR_UTF8_INVALID,
-                    $n::__Nonexhaustive => unreachable!("__Nonexhaustive value should not be instantiated"),
-                }
+#[derive(Debug, Eq, PartialEq)]
+pub enum Error {
+    HighLevel(HighLevelError),
+    LowLevel(LowLevelError),
+    HighAndLowLevel(HighLevelError, LowLevelError),
+    Other(c_int),
+    Utf8Error(Option<Utf8Error>),
+    // Stable-Rust equivalent of `#[non_exhaustive]` attribute. This
+    // value should never be used by users of this crate!
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+impl Error {
+
+    pub fn low_level(&self) -> Option<&LowLevelError> {
+        match &self {
+            &Error::LowLevel(error)
+            | &Error::HighAndLowLevel(_, error) => Some(&error),
+            _ => None
+        }
+    }
+
+    pub fn high_level(&self) -> Option<&HighLevelError> {
+        match &self {
+            &Error::HighLevel(error)
+            | &Error::HighAndLowLevel(error, _) => Some(&error),
+            _ => None
+        }
+    }
+
+    pub fn as_str(&self) -> String {
```

Comment:
Let's keep this as returning `&'static str`. For HighAndLowLevel you can just return the high level string. The `fmt::Display` implementation below shows the full error.